File: Sorting/bubble/bubble.py
# No imports are needed for this simple case

def bubble_sort(nums: list[int]) -> None:
   
    n = len(nums)
    # Traverse through all array elements
    for i in range(n - 1):
        swapped = False
        # Last i elements are already in place
        for j in range(0, n - i - 1):
            # Traverse the array from 0 to n-i-1
            # Swap if the element found is greater than the next element
            if nums[j] > nums[j + 1]:
                # This is the Pythonic way to swap two variables.
                # It uses tuple packing and unpacking.
                nums[j], nums[j + 1] = nums[j + 1], nums[j]
                swapped = True
        
        # If no two elements were swapped by inner loop, then the list is sorted
        if not swapped:
            break

def main():
   
    print("Enter numbers separated by spaces (e.g., 5 2 8 1):")
    line = input()

    #Parse the input string into a list of integers
    # We use a list comprehension for a clean, one-line conversion.
    # We also add a try-except block for robustness in case of bad input.
    try:
        numbers = [int(num_str) for num_str in line.split()]
    except ValueError:
        print("Invalid input. Please enter only integers separated by spaces.")
        return

    bubble_sort(numbers)

  
    print(f"Sorted Array: {numbers}")

# This is a standard Python construct to ensure main() runs only
# when the script is executed directly.
if __name__ == "__main__":
    main()

File: Sorting/bubble/bubble.go
package main

import (
	"bufio"   // For efficient reading
	"fmt"     // For printing
	"os"      // To access standard input
	"strconv" // To convert strings to integers
	"strings" // To split the input line
)

// bubbleSort sorts a slice of integers in-place.
// In Go, slices are reference types, so modifications inside
// the function will affect the original slice.
func bubbleSort(nums []int) {
	n := len(nums)
	for i := 0; i < n-1; i++ {
		swapped := false
		for j := 0; j < n-i-1; j++ {
			if nums[j] > nums[j+1] {
				// Go also allows for clean, one-line swaps.
				nums[j], nums[j+1] = nums[j+1], nums[j]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
}

func main() {
	// 1. Set up a reader to get input from the console
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("Enter numbers separated by spaces (e.g., 5 2 8 1):")
	
	// Read a single line, until the newline character
	line, _ := reader.ReadString('\n')
	// Trim whitespace (like the newline character) from the line
	line = strings.TrimSpace(line)

	// 2. Parse the input string into a slice of integers
	stringNumbers := strings.Split(line, " ")
	// Create a slice of integers with the correct size
	numbers := make([]int, 0, len(stringNumbers))

	for _, s := range stringNumbers {
		// strconv.Atoi returns the integer and an error
		num, err := strconv.Atoi(s)
		if err != nil {
			fmt.Println("Invalid input. Please enter only integers. Error:", err)
			return // Exit if there's an error
		}
		numbers = append(numbers, num)
	}

	// 3. Call the sorting function
	bubbleSort(numbers)

	// 4. Print the result
	// fmt.Println automatically formats slices in a readable way
	fmt.Println("Sorted Array:", numbers)
}